addons/sale_purchase/models/sale_order_line.py

```diff
@@ -159,99 +190,81 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
             uom_id=self.product_id.uom_po_id
         )
-        supplier_taxes = self.product_id.supplier_taxes_id.filtered(lambda t: t.company_id.id == self.company_id.id)
-        taxes = purchase_order.fiscal_position_id.map_tax(supplier_taxes)
 
-        # compute unit price
-        price_unit = 0.0
-        product_ctx = {
-            'lang': get_lang(self.env, purchase_order.partner_id.lang).code,
-            'company_id': purchase_order.company_id,
-        }
-        if supplierinfo:
-            price_unit = self.env['account.tax'].sudo()._fix_tax_included_price_company(
-                supplierinfo.price, supplier_taxes, taxes, self.company_id)
-            if purchase_order.currency_id and supplierinfo.currency_id != purchase_order.currency_id:
-                price_unit = supplierinfo.currency_id._convert(price_unit, purchase_order.currency_id, purchase_order.company_id, fields.Date.context_today(self))
-            product_ctx.update({'seller_id': supplierinfo.id})
-        else:
-            product_ctx.update({'partner_id': purchase_order.partner_id.id})
-
-        product = self.product_id.with_context(**product_ctx)
-        name = product.display_name
-        if product.description_purchase:
-            name += '\n' + product.description_purchase
+        price_unit, taxes = self._purchase_service_get_price_unit_and_taxes(supplierinfo, purchase_order)
+        name = self._purchase_service_get_product_name(supplierinfo, purchase_order, quantity)
 
         return {
             'name': name,
             'product_qty': purchase_qty_uom,
             'product_id': self.product_id.id,
             'product_uom': self.product_id.uom_po_id.id,
             'price_unit': price_unit,
-            'date_planned': fields.Date.from_string(purchase_order.date_order) + relativedelta(days=int(supplierinfo.delay)),
+            'date_planned': purchase_order.date_order + relativedelta(days=int(supplierinfo.delay)),
             'taxes_id': [(6, 0, taxes.ids)],
             'order_id': purchase_order.id,
             'sale_line_id': self.id,
         }
 
+    def _purchase_service_match_supplier(self, warning=True):
+        # determine vendor of the order (take the first matching company and product)
+        suppliers = self.product_id._select_seller(quantity=self.product_uom_qty, uom_id=self.product_uom)
+        if warning and not suppliers:
+            raise UserError(_("There is no vendor associated to the product %s. Please define a vendor for this product.") % (self.product_id.display_name,))
+        return suppliers
+
+    def _purchase_service_match_purchase_order(self, partner, company=False):
+        return self.env['purchase.order'].search([
+            ('partner_id', '=', partner.id),
+            ('state', '=', 'draft'),
+            ('company_id', '=', (company and company or self.env.company).id),
+        ], order='id desc')
```

Comment:
The extracted code did not specify a sorting order, so this could change the behavior.
It should be moved to a separate commit.
```suggestion
        ], limit=1)
```